### ISSUE TITLE
Make --training_samples actually slice the training dataset

### ISSUE THREAD

**Issue by msalganik** [niznik-dev/cruijff_kit#447]
## For Mattie

**Problem.** `setup_finetune.py` exposes a `--training_samples` flag (settable via `setup_finetune.yaml`), but it doesn't actually slice the dataset — today it only prints a warning if your step count looks low. The flag never reaches the generated `finetune.yaml`, and the dataset loader has no slice parameter to honor anyway. So setting `training_samples: 100` in a config does nothing to the actual fine-tune.

**Why fix it.** For consistency-curve experiments (vary N from 50 → 1000, eval on a fixed val set), we currently have to pre-generate 5 separate JSON files. With this fix we generate **one** large dataset once and slice with `training_samples` per run, while every run evaluates against the same large val set. Tighter, comparable CIs at small N; consistency tests get cheap to re-run.

**What changes after the fix.**
- `training_samples: 100` in `setup_finetune.yaml` → only the first 100 rows of `train` are loaded for fine-tuning that run.
- `dataset_val` is intentionally **unchanged** — every run sees the full val split, so val accuracy is comparable across N.
- Existing experiments that don't set `training_samples` are unaffected (no behavior change when the field is absent).
- Existing step-count guard / warnings still fire as before.
- Two new regression tests prevent this bug from returning, both in the standard `pytest tests/unit/` CI step.

**Acceptance.** Two fine-tune runs pointed at the same shared JSON, with `training_samples: 50` and `training_samples: 1000` in their respective `setup_finetune.yaml`s, produce `finetune.yaml`s that:
1. Differ only in `dataset.max_samples` (val sections identical).
2. Show step counts proportional to N in their SLURM logs (50-row run has roughly 1/20 the steps of the 1000-row run).
3. Both regression tests pass: `pytest tests/unit/test_setup_finetune_main.py tests/unit/test_chat_completion_dataset.py tests/unit/test_text_completion_dataset.py`.

**Effort estimate.** ~15 lines of source + ~70 lines of tests across 5 files. Small.

---

### Investigation already done

Don't re-investigate — go straight to the patch. Key findings from tracing the chain:

- `src/tools/torchtune/setup_finetune.py:20` — `META_ARGS = ["training_samples"]` excludes it from YAML output.
- `src/tools/torchtune/setup_finetune.py:739-741` — the args→YAML loop skips anything in `META_ARGS`, so `training_samples` is never written to `finetune.yaml`.
- `src/tools/torchtune/setup_finetune.py:683-691` — the only existing use of `training_samples` is the step-count guard. Keep this behavior.
- `src/tools/torchtune/datasets/chat_completion.py:113` — `chat_completion_dataset` factory loads all rows, no slice param. Same for `text_completion.py:99`.
- `src/tools/torchtune/templates/finetune_template.yaml:68-93` — template has separate `dataset` (split: train) and `dataset_val` (split: validation) blocks. Slice `dataset` only.

### Files to change

Source (3 files):
1. `src/tools/torchtune/datasets/chat_completion.py`
2. `src/tools/torchtune/datasets/text_completion.py`
3. `src/tools/torchtune/setup_finetune.py`

Tests (3 files: 1 modified, 2 new):
4. `tests/unit/test_setup_finetune_main.py` — add `TestTrainingSamples` class
5. `tests/unit/test_chat_completion_dataset.py` — new file
6. `tests/unit/test_text_completion_dataset.py` — new file (parallel to #5)

### Patch 1: `chat_completion.py`

In `chat_completion_dataset()` (~line 113), add a `max_samples: Optional[int] = None` kwarg (place near `max_length` in the signature). After the rows-resolution block (current lines ~155-160) and **before** `cfg = ChatCompletionConfig(...)`, insert:

```python
if max_samples is not None:
    rows = rows[:max_samples]
```

You do **not** need to add `max_samples` to `ChatCompletionConfig` — slicing at the factory level is sufficient because the dataset class consumes the resolved `rows` list. Update the docstring's example block to mention `max_samples` (one line).

### Patch 2: `text_completion.py`

Mirror the change in `text_completion_dataset()` (~line 99). Same kwarg, same slice location (between rows resolution and `cfg = TextCompletionConfig(...)`). Update docstring.

### Patch 3: `setup_finetune.py`

a. **Remove `"training_samples"` from `META_ARGS`** at line 20:
```python
META_ARGS = []  # leave the list for future meta-args
```

b. **Add a handler in the args→YAML loop** (the `for key, value in vars(args).items():` block starting at line 739). Place it among the dataset-related branches (around lines 779-796):
```python
elif key == "training_samples":
    if value is not None:
        config["dataset"]["max_samples"] = value
    # Intentionally NOT setting config["dataset_val"]["max_samples"]
    # — val set should stay full so consistency-curve runs are comparable across N.
```

c. **Update the help text** at line 593:
```python
help="Cap the number of rows loaded for training (slices dataset[:training_samples]). "
     "When provided, also computes total training steps and warns if dangerously low. "
     "Validation set is unaffected.",
```

d. **Leave the step-count guard at lines 683-691 unchanged.**

### Patch 4: `tests/unit/test_setup_finetune_main.py` — add regression test

Append a new class. The `run_main` and `setup_yaml` fixtures already exist in this file (defined at lines 16-61) — reuse them.

```python
class TestTrainingSamples:
    """Regression: --training_samples must slice train data, not val."""

    def test_training_samples_writes_max_samples_to_dataset(self, run_main):
        config, _ = run_main(extra_args=["--training_samples", "50"])
        assert config["dataset"]["max_samples"] == 50

    def test_training_samples_does_not_affect_dataset_val(self, run_main):
        config, _ = run_main(extra_args=["--training_samples", "50"])
        assert "max_samples" not in config.get("dataset_val", {}), \
            "training_samples must NOT slice the validation set — val must stay full across consistency-curve runs"

    def test_training_samples_via_config_file(self, run_main, setup_yaml):
        """Mirrors the consistency-curve workflow: knob set in setup_finetune.yaml, not CLI."""
        with open(setup_yaml) as f:
            config_data = yaml.safe_load(f)
        config_data["training_samples"] = 50
        with open(setup_yaml, "w") as f:
            yaml.dump(config_data, f)
        config, _ = run_main()
        assert config["dataset"]["max_samples"] == 50
        assert "max_samples" not in config.get("dataset_val", {})

    def test_no_training_samples_means_no_max_samples(self, run_main):
        config, _ = run_main()
        assert "max_samples" not in config["dataset"]
```

### Patch 5: `tests/unit/test_chat_completion_dataset.py` — new file

```python
"""Regression: chat_completion_dataset must honor max_samples for training-set slicing."""

import json
from unittest.mock import MagicMock, patch

import pytest

from cruijff_kit.tools.torchtune.datasets.chat_completion import chat_completion_dataset


@pytest.fixture
def small_json(tmp_path):
    path = tmp_path / "data.json"
    rows = {
        "train": [{"input": str(i), "output": str(i)} for i in range(10)],
        "validation": [{"input": "v", "output": "v"}],
    }
    path.write_text(json.dumps(rows))
    return path


@patch("transformers.AutoTokenizer.from_pretrained")
def test_max_samples_slices_train(mock_tok, small_json):
    mock_tok.return_value = MagicMock()
    ds = chat_completion_dataset(
        tokenizer=None, source="json", data_files=str(small_json),
        model_path="/fake/model", split="train", max_samples=3,
    )
    assert len(ds) == 3


@patch("transformers.AutoTokenizer.from_pretrained")
def test_max_samples_none_loads_all(mock_tok, small_json):
    mock_tok.return_value = MagicMock()
    ds = chat_completion_dataset(
        tokenizer=None, source="json", data_files=str(small_json),
        model_path="/fake/model", split="train",
    )
    assert len(ds) == 10


@patch("transformers.AutoTokenizer.from_pretrained")
def test_max_samples_larger_than_data_returns_full(mock_tok, small_json):
    """Python slice semantics: rows[:999] on a 10-row list returns 10 rows."""
    mock_tok.return_value = MagicMock()
    ds = chat_completion_dataset(
        tokenizer=None, source="json", data_files=str(small_json),
        model_path="/fake/model", split="train", max_samples=999,
    )
    assert len(ds) == 10
```

### Patch 6: `tests/unit/test_text_completion_dataset.py` — new file

Mirror Patch 5, importing `text_completion_dataset` from `cruijff_kit.tools.torchtune.datasets.text_completion`. The factory signature differs slightly (no `system_prompt`) — adjust the test calls accordingly. Same three tests.

### Test plan

Run all of these from the repo root with the `cruijff` conda env active.

**a. Unit tests (must all pass):**
```bash
pytest tests/unit/test_setup_finetune.py tests/unit/test_setup_finetune_main.py tests/unit/test_chat_completion_dataset.py tests/unit/test_text_completion_dataset.py -v
```

**b. setup_finetune.py integration check (manual verification):**
```bash
mkdir -p /tmp/training_samples_test && cd /tmp/training_samples_test
cp /scratch/gpfs/MSALGANIK/mjs3/ck-experiments/consistency_curve_parity_k10/Llama-3.2-3B-Instruct_N50/setup_finetune.yaml .
# add `training_samples: 50` to the YAML if not already present
python /home/mjs3/cruijff_kit/src/tools/torchtune/setup_finetune.py --config_file setup_finetune.yaml
grep -B 1 -A 12 "^dataset:" finetune.yaml         # expect: max_samples: 50
grep -B 1 -A 12 "^dataset_val:" finetune.yaml     # expect: NO max_samples key
```

**c. End-to-end smoke (one shared dataset, two slices):**
```bash
python -m cruijff_kit.tools.model_organisms.generate \
  --rule first --N 200 --input_type digits --k 10 --seed 42 \
  --design in_distribution \
  --output_dir /tmp/training_samples_test \
  --output shared.json
# Build two minimal experiment dirs, both data_files=/tmp/training_samples_test/shared.json,
# one with training_samples: 50, one with training_samples: 160.
# Submit both, verify the 50-row run reports a much smaller `Training step summary:` total
# than the 160-row run in their slurm-*.out logs.
```

**d. Lint:**
```bash
ruff check src/tools/torchtune/setup_finetune.py src/tools/torchtune/datasets/chat_completion.py src/tools/torchtune/datasets/text_completion.py tests/unit/test_setup_finetune_main.py tests/unit/test_chat_completion_dataset.py tests/unit/test_text_completion_dataset.py
ruff format --check src/tools/torchtune/setup_finetune.py src/tools/torchtune/datasets/chat_completion.py src/tools/torchtune/datasets/text_completion.py tests/unit/test_setup_finetune_main.py tests/unit/test_chat_completion_dataset.py tests/unit/test_text_completion_dataset.py
```

### PR conventions

- Branch: `<issue-number>-training-samples-slice`.
- PR body starts with `Closes #<issue-number>` (per `.github/pull_request_template.md`).
- Paste evidence into the Testing Instructions section: pytest output for (a), grep output for (b), and a step-count comparison from (c).

### Out of scope (do NOT do in this PR)

- Renaming `training_samples` → `max_train_samples` (breaks existing configs).
- Exposing val-set slicing.
- Auto-emitting `training_samples` from `scaffold-torchtune` from a `parameters.training_samples` field — keep manual-override-in-`setup_finetune.yaml` for now.

---

— Claude x Matt
